packages/react-query-next-experimental/src/ReactQueryStreamedHydration.tsx

```diff
@@ -0,0 +1,87 @@
+'use client'
+
+import type { DehydratedState, QueryClient } from '@tanstack/react-query'
+import { dehydrate, hydrate, useQueryClient } from '@tanstack/react-query'
+import * as React from 'react'
+import type { HydrationStreamProviderProps } from './HydrationStreamProvider'
+import { createHydrationStreamProvider } from './HydrationStreamProvider'
+
+const stream = createHydrationStreamProvider<DehydratedState>()
+
+/**
+ * This component is responsible for:
+ * - hydrating the query client on the server
+ * - dehydrating the query client on the server
+ */
+export function ReactQueryStreamedHydration(props: {
+  children: React.ReactNode
+  queryClient?: QueryClient
+  transformer?: HydrationStreamProviderProps<DehydratedState>['transformer']
+}) {
+  const queryClient = useQueryClient(props.queryClient)
+
+  // <server only>
+  const isSubscribed = React.useRef(false)
+  /**
+   * We need to track which queries were added/updated during the render
+   */
+  const [trackedKeys] = React.useState(() => new Set<string>())
```

Comment:
What happens if there are already queries in the cache, for example is some queries have been seeded/prefetched/are currently being prefetched (pending)? Are you then supposed to seed the browser RQ cache yourself too by having a `<Hydrate>` above these new providers in the tree (not possible for pending queries), or should this Set be initialised with any existing queries?

Reply:
I think they would be hydrated automatically, too. Given that the `ReactQueryStreamedHydration` is very high up your component tree, it would subscribe very early to the QueryCache, so it would also track those queries.

Reply:
What @TkDodo said, if you do something that changes the react query client cache, we'll subscribe to it and flush it next chance we get

Reply:
But what if you prime the cache before the rendering pass? Like if you await some data in a Server Component (maybe you need to generate metadata anyway), or you have some existing mechanism of prefetching that prefetches _before_ rendering?